### PR TITLE
Resolve "Convert Semaphores & Mutexes to Classes"

### LINKED ISSUE #240: Convert Semaphores & Mutexes to Classes
They're already part of the way there since the struct has a constructor. Might as well do it right.

### PR DESCRIPTION
**Pull request by Kfeavel**
Closes #240

### CODE REVIEW

**Comment by Kfeavel** [comment]
This should make @JamesOrson happier since I've realized he's right, we should be writing this kernel like it's C++, not C.

File: kernel/lib/semaphore.cpp [hunk 1]
```diff
@@ -18,117 +18,98 @@
...
-    } while(!COMPARE_EXCHANGE(sem, curVal, __ATOMIC_RELAXED));
+        // Fail using atomic relaxed because it may allow us to get to the "waiting" state faster.
+    } while (!COMPARE_EXCHANGE(curVal, __ATOMIC_RELAXED));
     // Return success
     return 0;
 }
 
-int sem_trywait(sem_t *sem) {
-    IS_SEMAPHORE_VALID(sem);
+int Semaphore::Trywait()
```
Comment (micahswitzer) [inline kernel/lib/semaphore.cpp]: Can this be `TryWait` since it's two separate words?

File: kernel/lib/semaphore.cpp [hunk 2]
```diff
@@ -18,117 +18,98 @@
...
-    } while(!COMPARE_EXCHANGE(sem, curVal, __ATOMIC_ACQUIRE));
+        // We need to fail on an Atomic Acquire Release because it will fail less often (i.e. fewer loop iterations)
+    } while (!COMPARE_EXCHANGE(curVal, __ATOMIC_ACQUIRE));
     // To get here the semaphore must not have been locked.
     return 0;
 }
 
-int sem_timedwait(sem_t *sem, const uint32_t *usec) {
-    IS_SEMAPHORE_VALID(sem);
+int Semaphore::Timedwait(const uint32_t* usec)
```
Comment (micahswitzer) [inline kernel/lib/semaphore.cpp]: Can this be `TimedWait` since it's two separate words?

File: kernel/lib/semaphore.cpp (hunk 1 above)
Comment (Kfeavel) [reply kernel/lib/semaphore.cpp]: Resolved in 30ae5702a50aceafa1b2bcee12b6f78f07d97fc7

File: kernel/lib/semaphore.cpp (hunk 2 above)
Comment (Kfeavel) [reply kernel/lib/semaphore.cpp]: Resolved in 30ae5702a50aceafa1b2bcee12b6f78f07d97fc7

**Review by micahswitzer — APPROVED** [review]

**Pull request merged**